File: src/big_tofu/create.py
from __future__ import annotations

from typing import Any

from big_config.utils import deep_merge, sort_nested_map
from big_tofu.core import Construct, add_suffix, caller_identity, construct, fqn_to_name, reference, root_arn


def bucket(fqn: str, *xs: str) -> list[Construct]:
    if xs:
        current = fqn
        for x in xs:
            current = add_suffix(current, f"-{x}")
        return bucket(current)
    return [Construct("resource", "aws_s3_bucket", fqn, [{"bucket": fqn_to_name(fqn, "-")}])]


def sqs(fqn: str) -> list[Construct]:
    return [Construct("resource", "aws_sqs_queue", fqn, {"name": fqn_to_name(fqn)})]


def kms(fqn: str) -> list[Construct]:
    kms_key = Construct("resource", "aws_kms_key", fqn, {})
    policy = Construct(
        "data",
        "aws_iam_policy_document",
        add_suffix(fqn, "-data-policy"),
        [
            {
                "statement": [
                    {
                        "actions": ["kms:*"],
                        "effect": "Allow",
                        "resources": ["*"],
                        "principals": [{"identifiers": [root_arn(caller_identity)], "type": "AWS"}],
                    }
                ]
            }
        ],
    )
    return [
        caller_identity,
        policy,
        kms_key,
        Construct(
            "resource",
            "aws_kms_key_policy",
            add_suffix(fqn, "-resource-policy"),
            {"key_id": reference(kms_key, "id"), "policy": reference(policy, "json")},
        ),
    ]


def provider(config: dict[str, Any]) -> dict[str, Any]:
    region = config.get("region")
    bucket_name = config.get("bucket")
    module = config.get("module")
    assume_role_value = config.get("assume-role", config.get("assume_role"))
    key = f"{str(module).lstrip(':').rsplit('/', 1)[-1]}.tfstate"
    assume_role = {"assume_role": {"role_arn": assume_role_value}} if assume_role_value and str(assume_role_value).strip() else {}
    return {
        "provider": {"aws": {"region": region, **assume_role}},
        "terraform": {
            "backend": {"s3": {"bucket": bucket_name, "encrypt": True, "key": key, "region": region, **assume_role}},
            "required_providers": {"aws": {"source": "hashicorp/aws", "version": "~> 5.0"}},
            "required_version": ">= 1.8.0",
        },
    }


__all__ = ["bucket", "sqs", "kms", "provider", "deep_merge", "sort_nested_map", "construct"]

File: src/big-tofu/create.ts
import { deepMerge } from "../utils.js";
import { nameOf, type Keyword, type Opts } from "../keys.js";
import { Construct, addSuffix, callerIdentity, construct, fqnToName, reference, rootArn } from "./core.js";

export function bucket(fqn: Keyword, ...xs: string[]): Construct[] {
  if (xs.length > 0) return bucket(addSuffix(fqn, `-${xs[0]}`), ...xs.slice(1));
  return [new Construct("resource", "aws_s3_bucket", fqn, [{ bucket: fqnToName(fqn, "-") }])];
}

export function sqs(fqn: Keyword): Construct[] {
  return [new Construct("resource", "aws_sqs_queue", fqn, { name: fqnToName(fqn) })];
}

export function kms(fqn: Keyword): Construct[] {
  const kmsKey = new Construct("resource", "aws_kms_key", fqn, {});
  const policy = new Construct("data", "aws_iam_policy_document", addSuffix(fqn, "-data-policy"), [{
    statement: [{
      actions: ["kms:*"],
      effect: "Allow",
      resources: ["*"],
      principals: [{ identifiers: [rootArn(callerIdentity)], type: "AWS" }]
    }]
  }]);
  return [
    callerIdentity,
    policy,
    kmsKey,
    new Construct("resource", "aws_kms_key_policy", addSuffix(fqn, "-resource-policy"), {
      key_id: reference(kmsKey, "id"),
      policy: reference(policy, "json")
    })
  ];
}

export function provider({ region, bucket, module, assumeRole, "assume-role": assumeRoleKebab }: { region: string; bucket: string; module: Keyword; assumeRole?: string; "assume-role"?: string }): Opts {
  const role = assumeRole ?? assumeRoleKebab;
  const key = `${nameOf(module)}.tfstate`;
  const assume_role = role && role.trim().length > 0 ? { assume_role: { role_arn: role } } : {};
  return {
    provider: { aws: { region, ...assume_role } },
    terraform: {
      backend: { s3: { bucket, encrypt: true, key, region, ...assume_role } },
      required_providers: { aws: { source: "hashicorp/aws", version: "~> 5.0" } },
      required_version: ">= 1.8.0"
    }
  };
}

export function constructsToObject(constructs: Construct[]): Opts {
  return deepMerge(...constructs.map((c) => construct(c)));
}

export { constructsToObject as "constructs->object" };
